forecasttools/utils.py

```diff
@@ -0,0 +1,183 @@
+"""
+A collection of utility functions used
+across other forecasttools code.
+"""
+
+from collections.abc import Iterable, MutableSequence
+from datetime import datetime, timedelta
+
+import arviz as az
+import polars as pl
+import xarray as xr
+
+
+def validate_input_type(
+    value: any, expected_type: type | tuple[type], param_name: str
+):
+    """Checks the type of a variable and
+    raises a TypeError if it does not match
+    the expected type."""
+    if isinstance(expected_type, tuple):
+        if not any(isinstance(value, t) for t in expected_type):
+            raise TypeError(
+                f"Parameter '{param_name}' must be one of the types {expected_type}; got {type(value)}"
+            )
+    else:
+        if not isinstance(value, expected_type):
+            raise TypeError(
+                f"Parameter '{param_name}' must be of type '{expected_type.__name__}'; got {type(value)}"
+            )
+
+
+def validate_and_get_start_time(start_date_iso: any):
+    """Handles the start_date_iso input,
+    converting it to a datetime object.
+
+    If the input is a string, it must be in
+    the format 'YYYY-MM-DD' or
+    'YYYY-MM-DD HH:MM:SS'. If it's a
+    datetime object, it is returned as is.
+    Otherwise, a TypeError is raised.
+    """
+    if isinstance(start_date_iso, str):
+        # try parsing both formats
+        for fmt in ("%Y-%m-%d", "%Y-%m-%d %H:%M:%S"):
+            try:
+                return datetime.strptime(start_date_iso, fmt)
+            except ValueError:
+                continue
+        # raise error if neither format matches
+        raise ValueError(
+            f"Parameter 'start_date_iso' must be in the format 'YYYY-MM-DD' or 'YYYY-MM-DD HH:MM:SS' if provided as a string; got {start_date_iso}."
+        )
+    elif isinstance(start_date_iso, datetime):
+        return start_date_iso
+    else:
+        raise TypeError(
+            f"Parameter 'start_date_iso' must be of type 'str' or 'datetime'; got {type(start_date_iso)}."
+        )
+
+
+def validate_and_get_idata_group(idata: az.InferenceData, group: str):
+    """Retrieves the group from the
+    InferenceData object and validates its
+    existence."""
+    idata_group = getattr(idata, group, None)
+    if idata_group is None:
+        raise ValueError(f"Group '{group}' not found in idata object.")
+    return idata_group
```

Comment:
I think I prefer 
```suggestion
    if not hasattr(idata, group):
        raise ValueError(f"Group '{group}' not found in idata object.")
    return getattr(idata, group)
```
this is also more consistent with the pattern in `validate_and_get_idata_group_var()` below